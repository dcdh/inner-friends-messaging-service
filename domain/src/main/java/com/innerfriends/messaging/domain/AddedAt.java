package com.innerfriends.messaging.domain;

import java.time.ZonedDateTime;
import java.util.Objects;

public final class AddedAt {

    private final ZonedDateTime at;

    public AddedAt(final EventAt eventAt) {
        this(eventAt.at());
    }

    public AddedAt(final PostedAt postedAt) {
        this(postedAt.at());
    }

    public AddedAt(final ZonedDateTime at) {
        this.at = at;
    }

    public ZonedDateTime at() {
        return at;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof AddedAt)) return false;
        AddedAt addedAt = (AddedAt) o;
        return Objects.equals(at, addedAt.at);
    }

    @Override
    public int hashCode() {
        return Objects.hash(at);
    }

    @Override
    public String toString() {
        return "AddedAt{" +
                "at=" + at +
                '}';
    }
}
