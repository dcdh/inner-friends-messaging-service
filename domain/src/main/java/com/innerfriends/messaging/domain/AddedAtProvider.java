package com.innerfriends.messaging.domain;

public interface AddedAtProvider {

    AddedAt generate();

}
