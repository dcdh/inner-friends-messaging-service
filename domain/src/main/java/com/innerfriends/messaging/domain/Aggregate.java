package com.innerfriends.messaging.domain;

import java.util.Objects;
import java.util.concurrent.Callable;
import java.util.function.Consumer;

public abstract class Aggregate {

    private Long version;

    public Aggregate(final Long version) {
        this.version = Objects.requireNonNull(version);
        if (version < 0) {
            throw new IllegalStateException();
        }
    }

    protected void apply(final Callable apply) {
        try {
            apply.call();
        } catch (final Exception exception) {
            throw new FailedToMutateAggregateException(exception);
        }
        version++;
    }

    public Long version() {
        return version;
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) return true;
        if (!(o instanceof Aggregate)) return false;
        final Aggregate aggregate = (Aggregate) o;
        return Objects.equals(version, aggregate.version);
    }

    @Override
    public int hashCode() {
        return Objects.hash(version);
    }

    @Override
    public String toString() {
        return "Aggregate{" +
                "version=" + version +
                '}';
    }
}
