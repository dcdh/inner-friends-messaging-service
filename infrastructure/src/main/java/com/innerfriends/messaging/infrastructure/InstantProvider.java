package com.innerfriends.messaging.infrastructure;

import javax.enterprise.context.ApplicationScoped;
import java.time.Instant;

@ApplicationScoped
public class InstantProvider {

    public Instant now() {
        return Instant.now();
    }

}
