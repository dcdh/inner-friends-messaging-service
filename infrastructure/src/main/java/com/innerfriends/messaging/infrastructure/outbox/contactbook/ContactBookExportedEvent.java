package com.innerfriends.messaging.infrastructure.outbox.contactbook;

import com.fasterxml.jackson.databind.JsonNode;
import io.debezium.outbox.quarkus.ExportedEvent;

public interface ContactBookExportedEvent extends ExportedEvent<String, JsonNode> {

    @Override
    default String getAggregateType() {
        return "ContactBook";
    }

}
