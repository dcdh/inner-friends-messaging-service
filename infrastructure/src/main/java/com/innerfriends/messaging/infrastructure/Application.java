package com.innerfriends.messaging.infrastructure;

import com.innerfriends.messaging.domain.*;
import com.innerfriends.messaging.domain.usecase.*;

import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.inject.Produces;
import java.time.ZonedDateTime;
import java.util.UUID;

public class Application {

    @Produces
    @ApplicationScoped
    public AddedAtProvider addedAtProviderProducer() {
        return () -> new AddedAt(ZonedDateTime.now());
    }

    @Produces
    @ApplicationScoped
    public ConversationIdentifierProvider conversationIdentifierProvider() {
        return new ConversationIdentifierProvider() {
            @Override
            public ConversationIdentifier generate(final OpenedBy openedBy) {
                return new ConversationIdentifier(openedBy.identifier().identifier() + "_" + UUID.randomUUID());
            }
        };
    }

    @Produces
    @ApplicationScoped
    public PostedAtProvider postedAtProvider() {
        return new PostedAtProvider() {
            @Override
            public PostedAt now() {
                return new PostedAt(ZonedDateTime.now());
            }
        };
    }

    @Produces
    @ApplicationScoped
    public AddContactIntoContactBookUseCase addContactIntoContactBookUseCaseProducer(final ContactBookRepository contactBookRepository,
                                                                                     final AddedAtProvider addedAtProvider) {
        return new AddContactIntoContactBookUseCase(contactBookRepository, addedAtProvider);
    }

    @Produces
    @ApplicationScoped
    public CreateContactBookUseCase createContactBookUseCaseProducer(final ContactBookRepository contactBookRepository) {
        return new CreateContactBookUseCase(contactBookRepository);
    }

    @Produces
    @ApplicationScoped
    public ListAllContactsUseCase listAllContactsUseCaseProducer(final ContactBookRepository contactBookRepository) {
        return new ListAllContactsUseCase(contactBookRepository);
    }

    @Produces
    @ApplicationScoped
    public ListConversationsUseCase listConversationsUseCaseProducer(final ConversationRepository conversationRepository) {
        return new ListConversationsUseCase(conversationRepository);
    }

    @Produces
    @ApplicationScoped
    public ListMessagesInConversationUseCase listMessagesInConversationUseCaseProducer(final ConversationRepository conversationRepository) {
        return new ListMessagesInConversationUseCase(conversationRepository);
    }

    @Produces
    @ApplicationScoped
    public ListRecentContactsUseCase listRecentContactsUseCaseProducer(final ContactBookRepository contactBookRepository) {
        return new ListRecentContactsUseCase(contactBookRepository);
    }

    @Produces
    @ApplicationScoped
    public PostNewMessageToConversationUseCase postNewMessageToConversationUseCase(final ConversationRepository conversationRepository,
                                                                                   final PostedAtProvider postedAtProvider) {
        return new PostNewMessageToConversationUseCase(conversationRepository, postedAtProvider);
    }

    @Produces
    @ApplicationScoped
    public OpenANewConversationUseCase startConversationUseCaseProducer(final ConversationRepository conversationRepository,
                                                                        final ContactBookRepository contactBookRepository,
                                                                        final ConversationIdentifierProvider conversationIdentifierProvider,
                                                                        final PostedAtProvider postedAtProvider) {
        return new OpenANewConversationUseCase(conversationRepository, contactBookRepository, conversationIdentifierProvider, postedAtProvider);
    }

}
