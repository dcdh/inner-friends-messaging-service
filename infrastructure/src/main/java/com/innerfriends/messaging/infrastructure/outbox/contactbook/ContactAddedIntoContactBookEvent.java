package com.innerfriends.messaging.infrastructure.outbox.contactbook;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.innerfriends.messaging.domain.Contact;
import com.innerfriends.messaging.domain.ContactBook;
import com.innerfriends.messaging.domain.Owner;
import com.innerfriends.messaging.infrastructure.InstantProvider;

import java.time.Instant;
import java.time.format.DateTimeFormatter;
import java.util.Objects;

public final class ContactAddedIntoContactBookEvent implements ContactBookExportedEvent {

    private final Owner owner;
    private final JsonNode addContactIntoContactBook;
    private final Instant timestamp;

    private ContactAddedIntoContactBookEvent(final Owner owner, final JsonNode addContactIntoContactBook, final Instant timestamp) {
        this.owner = Objects.requireNonNull(owner);
        this.addContactIntoContactBook = Objects.requireNonNull(addContactIntoContactBook);
        this.timestamp = Objects.requireNonNull(timestamp);
    }

    public static ContactAddedIntoContactBookEvent of(final ContactBook contactBookWithNewAddedContact,
                                                      final ObjectMapper objectMapper,
                                                      final InstantProvider instantProvider) {
        final Contact lastAddedContact = contactBookWithNewAddedContact.getLastAddedContact();
        final ObjectNode asJson = objectMapper.createObjectNode()
                .put("owner", contactBookWithNewAddedContact.owner().identifier().identifier())
                .put("version", contactBookWithNewAddedContact.version());
        final ObjectNode addedContact = asJson.putObject("addedContact");
        addedContact
                .put("contactIdentifier", lastAddedContact.contactIdentifier().identifier())
                .put("addedAt", lastAddedContact.addedAt().at().format(DateTimeFormatter.ISO_ZONED_DATE_TIME));
        return new ContactAddedIntoContactBookEvent(contactBookWithNewAddedContact.owner(), asJson, instantProvider.now());
    }

    @Override
    public String getAggregateId() {
        return owner.identifier().identifier();
    }

    @Override
    public String getType() {
        return "ContactAddedIntoContactBook";
    }

    @Override
    public Instant getTimestamp() {
        return timestamp;
    }

    @Override
    public JsonNode getPayload() {
        return addContactIntoContactBook;
    }
}
