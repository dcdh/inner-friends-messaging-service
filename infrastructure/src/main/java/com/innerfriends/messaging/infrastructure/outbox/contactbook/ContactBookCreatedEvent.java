package com.innerfriends.messaging.infrastructure.outbox.contactbook;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.innerfriends.messaging.domain.ContactBook;
import com.innerfriends.messaging.domain.Owner;
import com.innerfriends.messaging.infrastructure.InstantProvider;

import java.time.Instant;
import java.time.format.DateTimeFormatter;
import java.util.Objects;

public final class ContactBookCreatedEvent implements ContactBookExportedEvent {

    private final Owner owner;
    private final JsonNode addContactIntoContactBook;
    private final Instant timestamp;

    private ContactBookCreatedEvent(final Owner owner, final JsonNode addContactIntoContactBook, final Instant timestamp) {
        this.owner = Objects.requireNonNull(owner);
        this.addContactIntoContactBook = Objects.requireNonNull(addContactIntoContactBook);
        this.timestamp = Objects.requireNonNull(timestamp);
    }

    public static ContactBookCreatedEvent of(final ContactBook contactBookNewlyCreated,
                                             final ObjectMapper objectMapper,
                                             final InstantProvider instantProvider) {
        final ObjectNode asJson = objectMapper.createObjectNode()
                .put("owner", contactBookNewlyCreated.owner().identifier().identifier())
                .put("createdAt", contactBookNewlyCreated.createdAt().at().format(DateTimeFormatter.ISO_ZONED_DATE_TIME))
                .put("version", contactBookNewlyCreated.version());
        final ArrayNode contactsAsJson = asJson.putArray("contacts");
        contactBookNewlyCreated.allContacts().stream()
                .forEach(contact -> contactsAsJson.add(objectMapper.createObjectNode()
                        .put("contactIdentifier", contact.contactIdentifier().identifier())
                        .put("addedAt", contact.addedAt().at().format(DateTimeFormatter.ISO_ZONED_DATE_TIME)))
                );
        return new ContactBookCreatedEvent(contactBookNewlyCreated.owner(), asJson, instantProvider.now());
    }

    @Override
    public String getAggregateId() {
        return owner.identifier().identifier();
    }

    @Override
    public String getType() {
        return "ContactBookCreated";
    }

    @Override
    public Instant getTimestamp() {
        return timestamp;
    }

    @Override
    public JsonNode getPayload() {
        return addContactIntoContactBook;
    }

}
