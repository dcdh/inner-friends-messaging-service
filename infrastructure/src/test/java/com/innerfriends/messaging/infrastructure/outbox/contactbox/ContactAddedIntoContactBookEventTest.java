package com.innerfriends.messaging.infrastructure.outbox.contactbox;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.innerfriends.messaging.domain.*;
import com.innerfriends.messaging.infrastructure.InstantProvider;
import com.innerfriends.messaging.infrastructure.outbox.contactbook.ContactAddedIntoContactBookEvent;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;

import java.time.Instant;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.List;

import static org.approvaltests.Approvals.verifyJson;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;

@ExtendWith(MockitoExtension.class)
public class ContactAddedIntoContactBookEventTest {

    private InstantProvider instantProvider;

    private static final ObjectMapper objectMapper = new ObjectMapper();

    @BeforeEach
    public void setup() {
        this.instantProvider = mock(InstantProvider.class);
    }

    @Test
    public void should_return_expected_event() {
        // Given
        final Owner owner = new Owner("Mario");
        final ContactBook contactBook = new ContactBook(owner, buildCreatedAt(),
                List.of(new Contact(new ContactIdentifier("DamDamDeo"), new AddedAt(buildAddedAt(1).at()))));
        contactBook.addNewContact(new ContactIdentifier("Luigi"), buildAddedAt(1));
        contactBook.addNewContact(new ContactIdentifier("Peach"), buildAddedAt(2));
        doReturn(Instant.ofEpochSecond(1)).when(instantProvider).now();

        // When
        final ContactAddedIntoContactBookEvent contactAddedIntoContactBookEvent =
                ContactAddedIntoContactBookEvent.of(contactBook, objectMapper, instantProvider);

        // Then
        assertThat(contactAddedIntoContactBookEvent.getAggregateId()).isEqualTo("Mario");
        assertThat(contactAddedIntoContactBookEvent.getAggregateType()).isEqualTo("ContactBook");
        assertThat(contactAddedIntoContactBookEvent.getTimestamp()).isEqualTo(Instant.ofEpochSecond(1));
        assertThat(contactAddedIntoContactBookEvent.getType()).isEqualTo("ContactAddedIntoContactBook");
        verifyJson(contactAddedIntoContactBookEvent.getPayload().toString());
    }

    private CreatedAt buildCreatedAt() {
        return new CreatedAt(
                ZonedDateTime.of(2021, 10, 1, 0, 0, 0, 0, ZoneId.of("Europe/Paris")));
    }

    private AddedAt buildAddedAt(final Integer day) {
        return new AddedAt(ZonedDateTime.of(2021, 10, day, 0, 0, 0, 0, ZoneId.of("Europe/Paris")));
    }

}
