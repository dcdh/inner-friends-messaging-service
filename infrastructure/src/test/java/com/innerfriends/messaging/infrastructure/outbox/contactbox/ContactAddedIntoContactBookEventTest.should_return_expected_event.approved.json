{
  "owner": "Mario",
  "version": 2,
  "addedContact": {
    "contactIdentifier": "Peach",
    "addedAt": "2021-10-02T00:00:00+02:00[Europe/Paris]"
  }
}